### ISSUE TITLE
Add option to not show image hover for imprecise tokens

### ISSUE THREAD

**Issue by dineshm72** [Eriku33/Foundry-VTT-Image-Hover#69]
At some point in v11, foundry began to support the concept of imprecise tokens, at least in 5e, possibly in core.

Essentially, these are tokens that are not visible (they may be in darkness, may be invisible, etc.) but may be detected by other "imprecise" senses (hearing, tremorsense).

In game these tokens can be configured to display differently - common options are the mystery man, or (when using Vision 5e) a swirly soundwave image (see below).  The intention is that you know something is there, but you cannot identify it, friend or foe.  The Token Disposition colorization is also hidden, along with any nameplates or health bar indicators.
![image](https://github.com/user-attachments/assets/cbf11eab-1535-49d6-b10f-12c1b8161dbb)
![image](https://github.com/user-attachments/assets/633e67f8-4418-4f17-9d73-c2c5a976f8b2)
![image](https://github.com/user-attachments/assets/3fe17397-6780-4d5b-ba7a-ae7e9067dc22)

Unfortunately, Image Hover does not (yet!) respect this setting, and will gladly show the actor image in the designated location, kinda ruining this imprecision.
![image](https://github.com/user-attachments/assets/a4f039d4-ec10-4dae-beb8-6d3bd53f9cf0)

So the request is to add in handling for this case to (optionally?) prevent the image hover from working for these tokens.  Or perhaps only preventing it for players?  It may be useful to see for GMs I suppose.

Eber (foundry dev and author of Vision 5e) suggested that you need to check the token properties using something along these lines:

if (this.object.detectionLevel === Token.DETECTION_LEVELS.IMPRECISE ) ...

Thanks for considering!

**Comment by LukasPrism** [comment]
I've been loving this module for years in dnd5e and started playing a new campaign in pf2e with this very beloved module – unfortunately in our first session this module revealed portraits for creatures which are supposed to be hidden (they have an imprecise effect on them as mentioned by OP but the players shouldn't be able to tell what they are.

Is there any chance this will be addressed in the near future? In the meantime I think I'll need to disable this module.

**Comment by Eriku33** [comment]
Hello, feel free to disable the module for now, I will for sure make this change in the future

**Comment by dmaneyapanda2** [comment]
Well, out of the box I think 5e supports it.  But the base game doesn't actually do anything with that imprecise info I don't think.  

So a token which has the imprecise detection level just doesn't appear in the base game.  e.g. this is a token with no darkvision standing in the dark next to another token. 

![Image](https://github.com/user-attachments/assets/091eb1f4-873b-49db-9a57-066d8e0c68dd)

![Image](https://github.com/user-attachments/assets/486617bc-ae1e-4327-aaf8-7ff4e286fef7)

The easiest way to see it in action is to:
1. Install the Vision 5e module
2. Put on 2 tokens 10ft apart with normalish stats (because Vision 5e uses one of the stats to determine how far out the imprecise detection extends by default)
3. Make sure one has vision enabled, and no darkvision

When clicking on that token, the second token should now appear with the sound waves.

![Image](https://github.com/user-attachments/assets/a93b70f9-beb8-4ab0-aa0a-cd4f04bd74d9)

**Comment by dmaneyapanda2** [comment]
Though just to make sure you've seen it above, dev# suggested you probably just need to incorporate this sort of logic in the pipeline before showing the image:

if (this.object.detectionLevel === Token.DETECTION_LEVELS.IMPRECISE )

That was probably back in v11 or maybe v12.  Not sure if anything has changed since in 12 or 13.

**Comment by Eriku33** [comment]
Hello :),

With the help of @farling42, I have released a new version. Please try it out and let me know of any issues.

Thanks ❤️

**Comment by dmaneyapanda2** [comment]
It doesn't seem to be working for me in 5e 3.3.1.

![Image](https://github.com/user-attachments/assets/bcae90e0-400e-4c32-8b20-8e5f5f1a361e)

From the code it looks like there is supposed to be a setting for it, I don't see the setting either.

![Image](https://github.com/user-attachments/assets/a72b44c6-5133-464f-adb1-b93e15dc48b4)

**Comment by dmaneyapanda2** [comment]
Actually maybe you removed that setting and just rely on token.detectionFilter?  I dunno, either way it doesn't seem to work in 5e 3.3.

**Comment by farling42** [comment]
The settings option got removed in a more recent update. Now it always just ignores any token with a detectionFilter set on it.

**Comment by Eriku33** [comment]
Yes i removed the option since it's more of a bug that shouldn't exist instead of a feature.
I must be missing something, 
https://gyazo.com/14ba2cbfdad175a63f626f41de66d058

**Comment by dmaneyapanda2** [comment]
You are in dnd 4.3, I am on an older version 3.3.1.  Perhaps it's only supported in that newer version.

For me, I will be updating in the nearish future so if it does indeed work there I will be all set.  But I think a number of people will remain on 3.3 for a while, so it would be nice to find a solution for both.  But I'll leave that for you to decide.

**Comment by dmaneyapanda2** [comment]
Based on the code I linked above and my simple tests, if you test whether the token.detectionLevel != 1 before showing, that should work.

**Comment by farling42** [comment]
In dnd5e v4.3.6:
I made a creature token invisible.
I gave a PC token a detection mode of "feel tremor".

Selecting the PC token makes the invisible token be displayed with the swirly effect.
Hovering over the invisible token does not show an image from image hover.

**Comment by farling42** [comment]
> Based on the code I linked above and my simple tests, if you test whether the token.detectionLevel != 1 before showing, that should work.

`token.detectionLevel` doesn't exist in PF2

**Comment by Eriku33** [comment]
Just tested it with dnd5e v3.3.1 and still can't seem to reproduce it, just want to double-check that you are using image-hover v3.0.5?

**Comment by dmaneyapanda2** [comment]
I made a stealth edit above, it should be checking for ===2 not != 1, but that should be mostly the same I think, I'm not sure what if anything gets a value of 0 which was DETECTION_LEVELS_NONE in 3.3.  1 is PRECISE and 2 is IMPRECISE.

**Comment by dmaneyapanda2** [comment]
Hmm, maybe something in my game or other mods is conflicting, let me test with just IH again, sorry about the wild goose chase if this is a me problem!

**Comment by farling42** [comment]
I just installed Vision 5E, and that module is changing the system so that the "feel tremor" doesn't show the invisible token with the swirly filter on it.

**Comment by dmaneyapanda2** [comment]
Thanks guys.  With just IH (and V5E) you are right, it is working as expected, thanks!

It turns out that another module, Walled Templates, seems to be doing something that is making detectionFilters return null or undefined for every token.  I'll try to track that down with its author.

Thanks again!

**Comment by dmaneyapanda2** [comment]
Hmm, no, now I am seeing detectionFilters on the imprecise token but it's still showing the image hover for it too.

Gah, I may need to table this for now and come back to it fresh.

**Comment by Eriku33** [comment]
All good, i'm sure if you have this problem then others will too.  I would recommended creating a new world and only have 1 or 2 relevant modules active

**Comment by dmaneyapanda2** [comment]
OK, it seems to be a multi-module conflict, though one I don't really understand.  It requires having all three of Image Hover, Vision 5e, and Walled Templates.

Just IH plus a few libraries (works fine, it is showing the tremorsense imprecision on this token) - note that you also get the same result with IH and WT only:

![Image](https://github.com/user-attachments/assets/3f12aeb4-f9bd-4f15-8156-045450f6d2d2)

Just IH and V5E (the tremorsense effect is being replaced by the vision 5e swirl):

![Image](https://github.com/user-attachments/assets/e200f464-a1d1-4574-b044-b5819142661b)

IG, V5E, and WT (now the image is getting shown):

![Image](https://github.com/user-attachments/assets/9af69b6d-bd2f-4a36-9856-44a53ecf7e5c)

The really weird thing is that in the console I see a filter on this token, so as far as I can tell this should still be hitting that part of your code that would return out of displaying the image, but it's obviously not.  

![Image](https://github.com/user-attachments/assets/e4d948b9-85c2-4b91-936d-07597aa52c82)

![Image](https://github.com/user-attachments/assets/8298fc31-b54e-45c6-8702-b1cfabeaf8f4)

Unfortunately I'm not capable enough with a JS debugger to try to figure out what is going on there, and it's also somewhat complicated by me hosting on the forge so I can't easily make edits to the code to add in console logging.  I'll maybe post in the WT (or V5E) discords to see if they have some idea, but if not I'm not sure what else I can do on my end.

**Comment by LukasPrism** [comment]
I can confirm it's working great in PF2E, thanks so much for this update!

**Comment by farling42** [comment]
I've done some investigation, and it is definitely an issue that Walled Templates messes with the token in some way and provides it's own display of vague detection.

So that is more of a problem for Walled Template to fix than for this module to support it.
